### DISCUSSION TITLE
Loading a skeleton into Sleap

### DISCUSSION THREAD

**Discussion by racenico** [talmolab/sleap#1898] [Help!] [answered]
Hello,

I just started using Sleap in my study project to train a model for pose-estimation in mice.
I wanted to run this on my Laptop (Windows 11) so that I can work on it outside the laboratory but I ran into a problem setting the project up on my Laptop. The GUI and everything loads normally and I can open Sleap on my Laptop and load Videos and Frames from my Laptop. I just have a problem wih creating and loading the skeleton. If I want to create a node or an edge or just try to load in a finished skeleton nothing happens though sometimes I get a Pop-Up asking me if I want to replace existing nodes.
I tried installing Sleap again to exclude any installation mistakes but that didn't solve my issue. I am now sadly out of ideas how to fix my problem.
The only other thing I found is an error message that appears in the console everytime I try to add nodes or a skeleton:

```python
Traceback (most recent call last):
  File "C:\Studienprojekt\envs\sleap\lib\site-packages\sleap\gui\commands.py", line 441, in openSkeleton
    self.execute(OpenSkeleton)
  File "C:\Studienprojekt\envs\sleap\lib\site-packages\sleap\gui\commands.py", line 242, in execute
    command().execute(context=self, params=kwargs)
  File "C:\Studienprojekt\envs\sleap\lib\site-packages\sleap\gui\commands.py", line 138, in execute
    self.do_with_signal(context, params)
  File "C:\Studienprojekt\envs\sleap\lib\site-packages\sleap\gui\commands.py", line 164, in do_with_signal
    context.signal_update(cls.topics)
  File "C:\Studienprojekt\envs\sleap\lib\site-packages\sleap\gui\commands.py", line 215, in signal_update
    self.update_callback(what)
  File "C:\Studienprojekt\envs\sleap\lib\site-packages\sleap\gui\app.py", line 1169, in on_data_update
    self.skeleton_dock.nodes_table.model().items = self.state["skeleton"]
  File "C:\Studienprojekt\envs\sleap\lib\site-packages\sleap\gui\dataviews.py", line 103, in items
    item_data = self.item_to_data(obj, item)
  File "C:\Studienprojekt\envs\sleap\lib\site-packages\sleap\gui\dataviews.py", line 405, in item_to_data
    return dict(name=item.name, symmetry=obj.get_symmetry_name(item.name))
  File "C:\Studienprojekt\envs\sleap\lib\site-packages\sleap\skeleton.py", line 798, in get_symmetry_name
    symmetric_node = self.get_symmetry(node)
  File "C:\Studienprojekt\envs\sleap\lib\site-packages\sleap\skeleton.py", line 774, in get_symmetry
    node_node = self.find_node(node)
  File "C:\Studienprojekt\envs\sleap\lib\site-packages\sleap\skeleton.py", line 573, in find_node
    raise ValueError("Found multiple nodes named ({}).".format(name))
ValueError: Found multiple nodes named (name).
```

Can someone please tell me if they ran into the same problem in the past or know how to fix this?
Many thanks,
Cedric

**Comment by eberrigan** [comment]
Hi @racenico,

It seems like you have multiple nodes with the same name. Is that the case? Please let us know what your skeleton looks like the in the skeletons tab in the GUI.

Here are the lines in the code this refers to https://github.com/talmolab/sleap/blob/076f3dda2036c5833dec35ff4479d3ae1c7f7f55/sleap/skeleton.py#L555C1-L575C20. It does seem like the outdated `.format(name)` is not working and should be replaced with an F-string in the code. 

How familiar are you with Python? You may find this notebook helpful for debugging your issue https://colab.research.google.com/drive/1OmnmZIM64pWFqoeNkUm2QnpOWmPwoUbN?usp=sharing. Namely something like
```
import sleap
labels = sleap.load_file("<path/to/your/sleap/file.slp>")
print(labels.skeletons) # To check skeletons (there should only be one per project)
print(labels.skeleton) # Hopefully this output is the same as labels.skeletons in the case where there is one skeleton
print(labels.skeletons.nodes # Check the nodes in the skeleton
```

Best,

Elizabeth

**Reply by racenico** [reply, marked as answer]
Hello @eberrigan,

thanks for the reply.
My problem with the skeleton is that I couldn't even add a New Node in Sleap and always got that error It didn't appear in the skeletons tab. Also the skeleton is working in the Sleap version of my supervisor.

I replaced .format(name) in the file with an f-string but that didn't change anything. 

I am not really that proficient with python and have some difficulties loading my sleap-project file in the notebook.

I also just tried adjusting the version of my python and the packages used in sleap because I realised there apparently are known issues with tensorflow. I am sorry that I overlooked that before posting my question here.
I now reached a point where I can add Nodes and Edges and create a skeleton but I am still not able to load the finished skeleton my supervisor sent me although it is working for him and I still get the same error message and the skeleton doesn't appear in Sleap.
I recreated the skeleton my supervisor sent me and if he is fine with that I would consider my issue fixed. 

Thanks for the help,
Cedric

**Comment by eberrigan** [comment]
@racenico,

I see so everything is working for your supervisor? Make sure you have duplicates of your project so you don't mess something up and we recommend saving versions of your file to track changes made, especially to skeletons. This makes it much easier to backtrack if any mistakes are made. 

It sounds like a fresh install of the latest version of SLEAP in an isolated conda environment might help. Instructions are here depending on your OS: https://github.com/talmolab/sleap/releases/tag/v1.3.3.

Best,

Elizabeth

**Comment by roomrys** [comment]
In the following discussion:
- #1918

we found that the underlying problem comes from an upgrade to `attr>=22.2.0` where the `__getstate__` method of `attr` classes is changed to return a `dict` instead of a `tuple`. Then, when encoding the `Skeleton` using `jsonpickle`, `jsonpickle` instead encodes the `"py/state"` as a `dict`. 

This is not a problem in itself IF staying on the same/compatible version of `attr`, but when attempting to use this newly encoded skeleton.json in a different SLEAP installation that perhaps resolved to an older `attr<22.2.0`, the `Skeleton` is read in incorrectly, taking the dictionary `"py/state"` keys as the attribute values (instead of the attribute names). Hence the error we see with `Found multiple nodes named (name)` since the `Skeleton` quite literally was reconstructed with multiple nodes all named `name`.
